text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,286 @@
+traces, how to propagate context, and how to enrich traces with attributes.
+
+### Context propagation
+
+A message may pass many different components and layers in one or more
+intermediaries when it is propagated from the producer to the consumer. It
+cannot be assumed, and in many cases it is not even desired, that all those
+components and layers are instrumented and propagate context according to
+OpenTelemetry requirements.
+
+A _creation context_ allows correlating the producer with the consumers of a
+message, regardless of intermediary instrumentation. The creation context is
+created by the producer and must be propagated to the consumers. It must not be
+altered by intermediaries.  This context helps to model dependencies between
+producers and consumers, regardless of the underlying messaging transport
+mechanism and its instrumentation.
+
+Instrumentors are required to instrument producer and consumer applications
+so that context is attached to messages and extraced from messages in a
+coordinated way. Future versions of these conventions might recommend [context propagation according to certain industry standards](#standards-for-context-propagation).
+
+> A producer SHOULD attach a creation context to each message. The creation context
+> SHOULD be attached in a way so that it is not changed by intermediaries.
+
+### Trace structure, names, and attributes
+
+#### Producer
+
+Producers are responsible for attaching a creation context to a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus having a unique and distinct
+context per message is recommended, but not required.
+
+For each producer scenario, a "Publish" span needs to be created. This span
+measures the duration of the call or operation that provides messages for
+sending or publishing to an intermediary. This call or operation (and the
+related "Publish" span) can either refer to a single message or a batch of
+multiple messages.
+
+It is recommended to create a "Create" span for every single message. "Create"
+spans can be created during the "Publish" operation as children of the
+"Publish" span. Alternatively, "Create" spans can be created independently of
+the "Publish" operation. In that case, SDKs may provide mechanisms to allow
+attaching independent contexts with messages.
+
+If a "Create" span exists for a message, its context must be attached to the
+message. If no "Create" span exists for a message, the context of the related
+"Publish" span must be attached to the message.
```

Comment:
Based on the [flow](https://github.com/open-telemetry/oteps/pull/192/files#diff-a96f155dcb7d9ffc3712363edfab3d53d71b120724ca2cd968720c184a9ae0aaR69), this infers a preference of a "Create" context over a "Publish" context, even though "Publish" will often happen after "Create".

Is this is intentional?

If so, should the preference for "Create" context be explicitly called out?

At present it could be read as "Create" context being attached to message and then a subsequent "Publish" does the same, overwriting the "Create" context. As there doesn't appear to be anything stating what should be on the message when it reaches a Consumer in the case where the two contexts had been created on the Producer side.

Reply:
>  this infers a preference of a "Create" context over a "Publish" context, even though "Publish" will often happen after "Create".
>
> Is this is intentional?

Yes, the reason being that the "Create" context will always relate to a _single_ message, whereas a "Publish" context might relate to a batch of messages. If there's a context relating to a single message, it should be preferred.

> If so, should the preference for "Create" context be explicitly called out?

That's what we tried to do in this paragraph, if it's not clear enough we can improve on the wording:

> If a "Create" span exists for a message, its context must be attached to the message. If no "Create" span exists for a message, the context of the related "Publish" span must be attached to the message.